core/src/main/java/org/elasticsearch/cluster/metadata/IndexGraveyard.java

```diff
@@ -0,0 +1,466 @@
+        return MetaData.API_AND_GATEWAY;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        return (obj instanceof IndexGraveyard) && Objects.equals(tombstones, ((IndexGraveyard)obj).tombstones);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(tombstones);
+    }
+
+    /**
+     * Get the current unmodifiable index tombstone list.
+     */
+    public List<Tombstone> getTombstones() {
+        return tombstones;
+    }
+
+    @Override
+    public XContentBuilder toXContent(final XContentBuilder builder, final Params params) throws IOException {
+        builder.startArray(TOMBSTONES_FIELD.getPreferredName());
+        for (Tombstone tombstone : tombstones) {
+            tombstone.toXContent(builder, params);
+        }
+        return builder.endArray();
+    }
+
+    public IndexGraveyard fromXContent(final XContentParser parser) throws IOException {
+        return new IndexGraveyard(GRAVEYARD_PARSER.parse(parser));
+    }
+
+    @Override
+    public String toString() {
+        return "IndexGraveyard[" + tombstones + "]";
+    }
+
+    @Override
+    public void writeTo(final StreamOutput out) throws IOException {
+        out.writeVInt(tombstones.size());
+        for (Tombstone tombstone : tombstones) {
+            tombstone.writeTo(out);
+        }
+    }
+
+    @Override
+    public IndexGraveyard readFrom(final StreamInput in) throws IOException {
+        return new IndexGraveyard(in);
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public Diff<MetaData.Custom> diff(final MetaData.Custom previous) {
+        return new IndexGraveyardDiff((IndexGraveyard) previous, this);
+    }
+
+    @Override
+    public Diff<MetaData.Custom> readDiffFrom(final StreamInput in) throws IOException {
+        return new IndexGraveyardDiff(in);
+    }
+
+    public static IndexGraveyard.Builder builder() {
+        return new IndexGraveyard.Builder();
+    }
+
+    public static IndexGraveyard.Builder builder(final IndexGraveyard graveyard) {
+        return new IndexGraveyard.Builder(graveyard);
+    }
+
+    /**
+     * A class to build an IndexGraveyard.
+     */
+    final public static class Builder {
+        private List<Tombstone> tombstones;
+        private int numPurged = -1;
+
+        private Builder() {
+            tombstones = new ArrayList<>();
+        }
+
+        private Builder(IndexGraveyard that) {
+            tombstones = new ArrayList<>(that.getTombstones());
+        }
+
+        /**
+         * A copy of the current tombstones in the builder.
+         */
+        public List<Tombstone> tombstones() {
+            return Collections.unmodifiableList(tombstones);
+        }
+
+        /**
+         * Add a deleted index to the list of tombstones in the cluster state.
+         */
+        public Builder addTombstone(final Index index) {
+            tombstones.add(new Tombstone(index, System.currentTimeMillis()));
```

Comment:
I think it'd be cleaner if the `Builder` captured the current time once and just set all the tombstones for that builder to have the same time (rather than the tombstones for a single delete request that contains multiple indices having slightly different timestamps).